Add support for Tag Expressions

clrudolphi opened the pull request:
### 🤔 What's changed?

This PR adds support for tag expressions on step bindings via the `[Scope(Tag = "mytag or thatTag")]` syntax and via the attribute argument of a Hook binding, such as `[BeforeScenario("mytag or thatTag")]`.

Compatible with the VisualStudio extension.

### ⚡️ What's your motivation? 

Add tag-expression support bringing closer parity with the other Cucumber implementations.

### 🏷️ What kind of change is this?

- :zap: New feature (non-breaking change which adds new behaviour)


### ♻️ Anything particular you want feedback on?

Any corner cases overlooked?



### 📋 Checklist:

<!--- 
This is to help you remember all the little things we often forget to do!

Feel free to delete any tasks that are not relevant, or add new ones.
-->

- [X] I've changed the behaviour of the code
  - [X] I have added/updated tests to cover my changes.
- [X] My change requires a change to the documentation.
  - [X] I have updated the documentation accordingly.
- [x] Users should know about my change
  - [x] I have added an entry to the "[vNext]" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request & included my GitHub handle to the release contributors list.

----

*This text was originally taken from the [template of the Cucumber project](https://github.com/cucumber/.github/blob/main/.github/PULL_REQUEST_TEMPLATE.md), then edited by hand. [You can modify the template here.](https://github.com/reqnroll/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

clrudolphi committed:
Initial commit. This adds support for tag exprssions on Scopes assigned to bindings.

clrudolphi committed:
Fix an issue in which the BindingProviderService (as invoked OOP and via reflection by the VisualStudio Extension for Reqnroll) would provide a hardcoded prefix "@" to all Scope Tag values. This is no longer necessary with Tag Expression support.

clrudolphi committed:
Merge branch 'main' into Exploratory_TagExpression_Support

clrudolphi committed:
Draft update to the documentation on the topic of tag expressions

clrudolphi committed:
File scope namespace for the ReqnrollTagExpressionParser.

clrudolphi committed:
Merge branch 'Documentation_for_Tag_Expressions' into Exploratory_TagExpression_Support

clrudolphi committed:
Merge branch 'main' into Exploratory_TagExpression_Support

clrudolphi committed:
Added nuget package reference to Cucumber.TagExpressions; dropping the direct assembly reference.

clrudolphi committed:
Merge branch 'main' into Exploratory_TagExpression_Support

clrudolphi committed:
Update CHANGELOG.md

304NotModified committed:
Merge branch 'main' into Exploratory_TagExpression_Support

304NotModified committed:
Update CHANGELOG.md

gasparnagy committed:
code cleanup

gasparnagy committed:
Merge remote-tracking branch 'origin/main' into Exploratory_TagExpression_Support

gasparnagy commented on Reqnroll/Bindings/Discovery/BindingSourceProcessor.cs:
```diff
@@ -75,7 +78,7 @@ public virtual void BuildingCompleted()
         {
             return attributes.Where(attr => attr.AttributeType.TypeEquals(typeof(ScopeAttribute)))
-                .Select(attr => new BindingScope(attr.TryGetAttributeValue<string>("Tag"), attr.TryGetAttributeValue<string>("Feature"), attr.TryGetAttributeValue<string>("Scenario")));
+                .Select(attr => new BindingScope(_tagExpressionParser.Parse(attr.TryGetAttributeValue<string>("Tag")), attr.TryGetAttributeValue<string>("Feature"), attr.TryGetAttributeValue<string>("Scenario")));
```
How do we represent scopes without tag expression. Because `TryGetAttributeValue` here will return null, but the Parse method does not handle null or empty values as far as I have seen, so this will not work. If we want to represent not-specified tag expression with a `null`, we have to make the parse method handle that or have a special parse helper that does this for it.

clrudolphi replied:
The `TagExpressionParser` does handle null input and will return a `NullExpression`. That is already included in the test suite for the ``Cucumber.TagExpression` parser.
The `BindingScope` `Match` method skips processing of the tag expression if it is a `NullExpression`.

gasparnagy reviewed (changes requested):
I think this is going to be fine, I have made a few comments that require a bit of rework.

Thinking it over, we should also postpone this to v4, because it might break some tooling.

clrudolphi committed:
Changes per review comments:
Enforced use of '@' prefix on all tag expressions except single-term (presumably) legacy expressions.
Encapsulated Cucumber.TagExpressionParser inside a Reqnroll.TagExpressionParser to control lifetime of the parser object.

clrudolphi committed:
Update scoped-bindings.md

Update doc to explicitly state that the '@' prefix is required in tag expressions (except for legacy single-term expressions).

clrudolphi commented on docs/automation/scoped-bindings.md:
````diff
@@ -28,6 +28,23 @@ Use the `[Scope]` attribute to define the scope:
+|@gui or @database|	Scenarios tagged with either @gui or @database|
+
+```{note}
+The '@' prefix is optional within a tag expression.
````
Documentation updated.

gasparnagy committed:
code cleanup

gasparnagy committed:
fix BindingProviderService to work with empty Tag

clrudolphi commented on Reqnroll/Bindings/Discovery/BindingSourceProcessor.cs:
```diff
@@ -75,7 +78,7 @@ public virtual void BuildingCompleted()
         {
             return attributes.Where(attr => attr.AttributeType.TypeEquals(typeof(ScopeAttribute)))
-                .Select(attr => new BindingScope(attr.TryGetAttributeValue<string>("Tag"), attr.TryGetAttributeValue<string>("Feature"), attr.TryGetAttributeValue<string>("Scenario")));
+                .Select(attr => new BindingScope(_tagExpressionParser.Parse(attr.TryGetAttributeValue<string>("Tag")), attr.TryGetAttributeValue<string>("Feature"), attr.TryGetAttributeValue<string>("Scenario")));
```
I have an approach implemented in which the result of parsing a bad tag expression results in an instance of a new class, `InvalidTagExpression` : `ITagExpression`. The `BindingScope` has been enhanced to surface this expression as a property The `BindingSource` class has a new method: `ValidateBindingScope` which checks for this type of tag expression and if found adds an Error BindingValidationResult. 
`BindingSourceProcessor.ValidateStepDefinition` and `BindingSourceProcessor.ValidateHook` methods call upon the new `ValidateBindingScope` method.

I am getting strange results.
I have a test (calculator) project with the following additional (5th) binding method added:
```c#
    [Given("something"), Scope(Tag = "foo bar")]
    public void GivenAnInvalidScopeTagExpressionIsUsed()
    {
        // This step definition is intentionally left blank.
        // It serves to test the handling of invalid scope tag expressions.
    }
``` 

When a test project is run via `dotnet test`, I get the error messages expected:
```
   C:\Users\clrud\source\repos\scratch\TagExpressionErrors\TagExpressionErrors\Features\Calculator.feature(10): error TESTERROR:
      Add two numbers (47ms): Error Message: Test method TagExpressionErrors.Features.CalculatorFeature.AddTwoNumbers threw exception:
      Reqnroll.BindingException: Binding error(s) found:
      Tag expression "foo bar" could not be parsed because of syntax error: Expected operator. (at offset 4)
```

However, in VisualStudio the results don't match the user experience you demonstrated for a faulty regex binding expression.
Output from Reqnroll is:
<img width="859" height="125" alt="image" src="https://github.com/user-attachments/assets/066591dd-9eea-4725-b8ee-d89c1b2c1a4f" />
Note: it indicates 4 imported step definitions. The 5th one is the one with the bad tag expression and it only gives a warning. But, the warning message is not the text expected; I'm not sure where that is coming from.

BUT, the Error List window is empty:
<img width="1038" height="189" alt="image" src="https://github.com/user-attachments/assets/814da18c-7b1d-449b-a181-0def0ce58071" />

If I execute the test project via the VS Test Explorer, I get a test log that matches what the command line test execution gives.
So I suspect something with the VS Extension. I'm running 2025.3.395 (Reqnroll for Visual Studio 2022 and 2026).

Any suggestions?

clrudolphi replied:
Figured out the VS user experience I'm seeing. 

The fact that the error message shown is different than the one produced by the `Reqnroll.TagExpressionParser` is because it is produced by the tag expression parser already built in to the VS Extension.
In order to have those messages displayed in the Errors window will require changes to the Extension.

I'll work on a separate PR for the VS Extension re the error window display. Would you like me to replace the old parser with the new? Or, I could modify the existing so that the error messages it produces mimics those produced by the new one.

clrudolphi replied:
The tag expression parser is [here](https://github.com/reqnroll/Reqnroll.VisualStudio/tree/e803e973ddba108954a87988787191470217550e/Reqnroll.VisualStudio/Discovery/TagExpressions)

In [this line](https://github.com/reqnroll/Reqnroll.VisualStudio/blob/e803e973ddba108954a87988787191470217550e/Reqnroll.VisualStudio/Discovery/BindingImporter.cs#L38) of the `BindingImporter` the scope for a step definition is caused to be parsed [here](https://github.com/reqnroll/Reqnroll.VisualStudio/blob/e803e973ddba108954a87988787191470217550e/Reqnroll.VisualStudio/Discovery/BindingImporter.cs#L147) - which invokes the built-in `TagExpressionParser`. If the expression is invalid, the `TagExpressionParser` will throw an exception which results in the step definition binding being dropped (and the `BindingImporter.ImportStepDefinition()` method will return a null).

What needs to change, I suspect, is that we catch the exception from the tag expression parser and swallow it earlier and let the stepDefinition bubble up where its Failed `DiscoveryResult` value can be acted upon.

We would want to add a method similar to [this one](https://github.com/reqnroll/Reqnroll.VisualStudio/blob/e803e973ddba108954a87988787191470217550e/Reqnroll.VisualStudio/Discovery/DiscoveryInvoker.cs#L174) in the `DiscoveryInvoker`.

Another option to consider is to eliminate the tag expression parsing completely. Its already been validated by the `Reqnroll.Binding.BindingSourceProcessor` and the result is part of the `DiscoveryResult`.

clrudolphi committed:
Wrapping tag parsing errors in a new type of ITagExpression which drives the BindingSourceProcessor to add BindingErrors to the BindingResult.

clrudolphi committed:
Merge branch 'main' into Exploratory_TagExpression_Support

clrudolphi committed:
Updated tests.
Updated Changelog.
Added sample to Formatters tests that demonstrate use of tag expressions on hooks.

clrudolphi committed:
Added properties to surface tag expression errors via BindingSourceProcessor.
Added BindingSourceProcessor test for handling of bad Scope tag expressions.

clrudolphi committed:
Refactored tag expression support with addition of a ReqnollTagExpression that tracks the original tag text.
BindingScope now pulls its ErrorMessage directly from its input tagExpression.
Fixed errors in BindingSourceProcessor handling of scope errors on hooks.
Added tests to BindingSourceProcessorTests demonstrating handling of bad scope expressions.

clrudolphi committed:
Fix ReqnrollTagExpression missing ToString() override

clrudolphi committed:
Merge branch 'main' into Exploratory_TagExpression_Support

gasparnagy committed:
small fixes

gasparnagy reviewed (approved):
This is good to go. Let's wait for v4 time to open before merge.

clrudolphi commented:
@gasparnagy Something with the build script and test generator is messed up. The first step of each test creates a new solution, saved as a snlx. The next step fails when attempting to a csproj to the "sln" file.

gasparnagy committed:
Merge remote-tracking branch 'origin/main' into Exploratory_TagExpression_Support

clrudolphi commented:
> > ah. there is a real error as well...
> Looks like something to do with how a tag expression is canonicalized during .ToString().
> 
> I'll take a look.

Hmm. Don't know why I hadn't thought this through before. The error comes from a Formatters acceptance test that I specifically added because the CCK doesn't have any examples of how complex tag expressions are rendered in Messages. With the latest changes we had made, I had adjusted whether a Scope would render its tag in original text form or in canonical form.

I've put a question into the CCK discord. Hoping to prompt them to create a CCK scenario that will lock down the rule.

So, this should wait until we get some feedback returned from the Cucumber team.

clrudolphi committed:
Adjusted acceptance test file for tag-expressions Formatters test scenario.

gasparnagy commented:
> the CCK doe

Super. Thx for the fix.
